### DISCUSSION TITLE
Shallow routing not working as expected

### DISCUSSION THREAD

**Discussion by asimplify** [vercel/next.js#47983] [Help]
### Summary

I am trying to change the query params using router.push without reloading the page. I tried using shallow routing as suggested by nextjs docs but it is not working as expected. It should be noted that I am using dynamic routing. So my url looks like this

http://localhost:3000/category/865/940/940?parentId=940

and as it is dynamic, the slug url would look like this

http://localhost:3000/category/[category]/[subCategory]/[productCat]?parentId=940

Now when I try shallow routing, it adds the query param but my url becomes something like this:

http://localhost:3000/category/[category]/[subCategory]/[productCat]?parentId=940&price=333

It should be like this:

http://localhost:3000/category/865/940/940?parentId=940&price=333

I used shallow in accordance to nextjs docs such as: router.push("&price=sdds", undefined, { shallow: true,})

but it giving me a weird url and refreshes the page.

**Comment by icyJoseph** [comment]
Hi,

Could you try with:

```js
import { useRouter } from "next/router";

const Page = () => {
  const router = useRouter();

  const navigate = () => {
    const query = { ...router.query, price: Math.floor(Math.random() * 100) };
    
    router.push({ query, pathname: router.pathname }, undefined, {
      shallow: true,
    });
  };

  return (
    <div>
      <button onClick={navigate}>Click me</button>
    </div>
  );
};

export const getServerSideProps = () => {
  console.count("GSSP");

  return { props: {} };
};

export default Page;

```

I am not sure where you got the `router.push("&price=sdds", undefined, { shallow: true,})` bit, but it is working as expected, because passing a pathname with a leading `/` has a specific meaning, which is to change the URL path relative to its immediate parent, whereas `/&price=sdds` would change the route from the root, leading you to a 404.

What you want is either the pathname with the queries that contains the dynamic segments, or a complete path such as `/category/100/100/100`. I think the first approach is easier.

Note that in the docs:

![Screenshot 2023-04-06 at 09 42 02](https://user-images.githubusercontent.com/21013447/230308557-f4eace2d-dae1-41ad-b16f-a137e67d813c.png)

They use `/?counter=10`, and they state, current path is `/`.
